File: searx/engines/bing.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""This is the implementation of the Bing-Web engine.  Some of this
implementations are shared by other engines:

- :ref:`bing images engine`
- :ref:`bing news engine`
- :ref:`bing videos engine`

.. note::

   Some functionality (paging and time-range results) are not supported
   since they depend on JavaScript.
"""

import base64
import re
import typing as t
from urllib.parse import parse_qs, urlencode, urlparse

import babel
import babel.languages
from lxml import html

from searx.enginelib.traits import EngineTraits
from searx.locales import region_tag
from searx.utils import eval_xpath, eval_xpath_getindex, eval_xpath_list, extract_text

if t.TYPE_CHECKING:
    from searx.extended_types import SXNG_Response
    from searx.search.processors import OnlineParams

about: dict[str, t.Any] = {
    "website": "https://www.bing.com",
    "wikidata_id": "Q182496",
    "official_api_documentation": "https://github.com/MicrosoftDocs/bing-docs",
    "use_official_api": False,
    "require_api_key": False,
    "results": "HTML",
}

# engine dependent config
categories = ["general", "web"]
safesearch = True
_safesearch_map: dict[int, str] = {
    0: "off",
    1: "moderate",
    2: "strict",
}
"""Filter results. 0: None, 1: Moderate, 2: Strict"""

base_url = "https://www.bing.com/search"
"""Bing-Web search URL"""


def get_locale_params(engine_region: str | None) -> dict[str, str] | None:
    """API documentation states the ``mkt`` parameter is *the
    recommended primary signal* for locale:

        If known, you are encouraged to always specify the market.
        Specifying the market helps Bing route the request and return an
        appropriate and optimal response.

    The ``mkt`` parameter takes a full ``<language>-<country>`` code.

    This function is shared with :py:mod:`searx.engines.bing_images`,
    :py:mod:`searx.engines.bing_news`, and :py:mod:`searx.engines.bing_videos`.
    """

    if not engine_region or engine_region == "clear":
        return None

    return {"mkt": engine_region}


def override_accept_language(params: "OnlineParams", engine_region: str | None) -> None:
    """Override the ``Accept-Language`` header.

    The default header built by :py:class:`~searx.search.processors.online.OnlineProcessor`
    appends ``en;q=0.3`` as a fallback language::

        Accept-Language: de,de-DE;q=0.7,en;q=0.3

    Bing seems to better select the results locale based on the
    ``Accept-Language`` value header.

    This function is shared with :py:mod:`searx.engines.bing_images`,
    :py:mod:`searx.engines.bing_news`, and :py:mod:`searx.engines.bing_videos`.
    """

    if not engine_region or engine_region == "clear":
        return

    lang = engine_region.split("-")[0]
    params["headers"]["Accept-Language"] = f"{engine_region},{lang};q=0.9"


def request(query: str, params: "OnlineParams") -> "OnlineParams":
    """Assemble a Bing-Web request."""

    engine_region = traits.get_region(params["searxng_locale"], traits.all_locale)

    override_accept_language(params, engine_region)

    query_params: dict[str, str | int] = {
        "q": query,
        "adlt": _safesearch_map.get(params.get("safesearch", 0), "off"),
    }

    locale_params = get_locale_params(engine_region)
    if locale_params:
        query_params.update(locale_params)

    params["url"] = f"{base_url}?{urlencode(query_params)}"

    # in some regions where geoblocking is employed (e.g. China),
    # www.bing.com redirects to the regional version of Bing
    params["allow_redirects"] = True

    return params


def response(resp: "SXNG_Response") -> list[dict[str, t.Any]]:
    """Get response from Bing-Web"""

    results: list[dict[str, t.Any]] = []

    dom = html.fromstring(resp.text)

    for item in eval_xpath_list(dom, '//ol[@id="b_results"]/li[contains(@class, "b_algo")]'):
        link = eval_xpath_getindex(item, ".//h2/a", 0, None)
        if link is None:
            continue

        href = link.attrib.get("href", "")
        title = extract_text(link)

        if not href or not title:
            continue

        # what about cn.bing.com, ..?
        if href.startswith("https://www.bing.com/ck/a?"):
            qs = parse_qs(urlparse(href).query)
            u_values = qs.get("u")
            if u_values:
                u_val = u_values[0]
                if u_val.startswith("a1"):
                    encoded = u_val[2:]
                    # base64url without padding
                    encoded += "=" * (-len(encoded) % 4)
                    href = base64.urlsafe_b64decode(encoded).decode("utf-8", errors="replace")

        # remove decorative icons that Bing injects into <p> elements
        # (`<span class="algoSlug_icon">`)
        content_els = eval_xpath(item, ".//p")
        for p in content_els:
            for icon in p.xpath('.//span[@class="algoSlug_icon"]'):
                icon.getparent().remove(icon)
        content = extract_text(content_els)

        results.append({"url": href, "title": title, "content": content})

    if results:
        result_len_container = "".join(eval_xpath(dom, '//span[@class="sb_count"]//text()'))
        result_len_container = re.sub(r"[^0-9]", "", result_len_container)
        if result_len_container:
            results.append({"number_of_results": int(result_len_container)})

    return results


def fetch_traits(engine_traits: EngineTraits) -> None:
    """Fetch regions from Bing-Web."""
    # pylint: disable=import-outside-toplevel

    from searx.network import get  # see https://github.com/searxng/searxng/issues/762
    from searx.utils import gen_useragent

    headers = {
        "User-Agent": gen_useragent(),
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
        "Accept-Language": "en-US;q=0.5,en;q=0.3",
        "DNT": "1",
        "Connection": "keep-alive",
        "Upgrade-Insecure-Requests": "1",
        "Sec-GPC": "1",
        "Cache-Control": "max-age=0",
    }

    resp = get("https://www.bing.com/account/general", headers=headers, timeout=5)
    if not resp.ok:
        raise RuntimeError("Response from Bing is not OK.")

    dom = html.fromstring(resp.text)

    map_market_codes: dict[str, str] = {
        "zh-hk": "en-hk",  # not sure why, but at Microslop this is the market code for Hongkong
    }

    for href in eval_xpath(dom, '//div[@id="region-section-content"]//div[@class="regionItem"]/a/@href'):
        cc_tag = parse_qs(urlparse(href).query)["cc"][0]
        if cc_tag == "clear":
            engine_traits.all_locale = cc_tag
            continue

        # add market codes from official languages of the country ..
        for lang_tag in babel.languages.get_official_languages(cc_tag, de_facto=True):
            lang_tag = lang_tag.split("_")[0]  # zh_Hant --> zh
            market_code = f"{lang_tag}-{cc_tag}"  # zh-tw
            market_code = map_market_codes.get(market_code, market_code)

            try:
                sxng_tag = region_tag(babel.Locale.parse("%s_%s" % (lang_tag, cc_tag.upper())))
            except babel.UnknownLocaleError:
                # silently ignore unknown languages
                continue

            conflict = engine_traits.regions.get(sxng_tag)
            if conflict:
                if conflict != market_code:
                    print("CONFLICT: babel %s --> %s, %s" % (sxng_tag, conflict, market_code))
                continue

            engine_traits.regions[sxng_tag] = market_code

File: go/internal/engines/bing/bing.go
// Package bing implements the Bing web search engine.
//
// Port of searx/engines/bing.py. Parses HTML results from Bing's search page.
package bing

import (
	"encoding/base64"
	"fmt"
	"net/http"
	"net/url"
	"strings"

	"golang.org/x/net/html"

	"github.com/searxng/searxng-go/internal/config"
	"github.com/searxng/searxng-go/internal/engine"
	"github.com/searxng/searxng-go/internal/result"
)

var safeSearchMap = map[int]string{
	0: "off",
	1: "moderate",
	2: "strict",
}

type Bing struct {
	cfg config.EngineConfig
}

func New(cfg config.EngineConfig) (engine.Engine, error) {
	return &Bing{cfg: cfg}, nil
}

func (b *Bing) Name() string             { return b.cfg.Name }
func (b *Bing) Categories() []string      { return b.cfg.Categories }
func (b *Bing) Shortcut() string          { return b.cfg.Shortcut }
func (b *Bing) SupportsPaging() bool      { return false }
func (b *Bing) SupportsTimeRange() bool   { return false }
func (b *Bing) SupportsSafeSearch() bool  { return true }
func (b *Bing) MaxPage() int              { return 1 }

func (b *Bing) BuildRequest(query string, params *engine.SearchParams) (*http.Request, error) {
	q := url.Values{}
	q.Set("q", query)

	ss := safeSearchMap[params.SafeSearch]
	if ss == "" {
		ss = "off"
	}
	q.Set("adlt", ss)

	// Set market based on language.
	if params.Language != "" && params.Language != "auto" && params.Language != "all" {
		q.Set("mkt", params.Language)
	}

	reqURL := "https://www.bing.com/search?" + q.Encode()

	req, err := http.NewRequest("GET", reqURL, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
	req.Header.Set("Accept-Language", "en-US,en;q=0.5")
	return req, nil
}

func (b *Bing) ParseResponse(resp *http.Response, body []byte) ([]result.Result, error) {
	doc, err := html.Parse(strings.NewReader(string(body)))
	if err != nil {
		return nil, fmt.Errorf("parsing HTML: %w", err)
	}

	var results []result.Result

	// Find result items: <li class="b_algo"> inside <ol id="b_results">
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if isResultItem(n) {
			if r, ok := parseResultItem(n); ok {
				results = append(results, r)
			}
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(doc)

	return results, nil
}

// isResultItem checks if a node is <li class="b_algo">.
func isResultItem(n *html.Node) bool {
	if n.Type != html.ElementNode || n.Data != "li" {
		return false
	}
	class := getAttr(n, "class")
	return strings.Contains(class, "b_algo")
}

func parseResultItem(li *html.Node) (result.Result, bool) {
	// Find <h2><a href="...">title</a></h2>
	var linkNode *html.Node
	findLink(li, &linkNode)
	if linkNode == nil {
		return result.Result{}, false
	}

	href := getAttr(linkNode, "href")
	title := extractText(linkNode)
	if href == "" || title == "" {
		return result.Result{}, false
	}

	// Decode Bing redirect URLs: https://www.bing.com/ck/a?u=a1<base64url>
	href = decodeBingURL(href)

	// Find content in <p> elements.
	content := findParagraphText(li)

	return result.Result{
		URL:      href,
		Title:    title,
		Content:  content,
		Category: "general",
		Template: "default",
		Priority: "normal",
	}, true
}

// decodeBingURL decodes Bing's redirect tracking URLs.
// Matches the Python logic at bing.py:141-151.
func decodeBingURL(href string) string {
	if !strings.HasPrefix(href, "https://www.bing.com/ck/a?") {
		return href
	}
	u, err := url.Parse(href)
	if err != nil {
		return href
	}
	uVal := u.Query().Get("u")
	if uVal == "" || !strings.HasPrefix(uVal, "a1") {
		return href
	}
	encoded := uVal[2:]
	// base64url without padding
	if m := len(encoded) % 4; m != 0 {
		encoded += strings.Repeat("=", 4-m)
	}
	decoded, err := base64.URLEncoding.DecodeString(encoded)
	if err != nil {
		return href
	}
	return string(decoded)
}

func findLink(n *html.Node, result **html.Node) {
	if n.Type == html.ElementNode && n.Data == "h2" {
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			if c.Type == html.ElementNode && c.Data == "a" {
				*result = c
				return
			}
		}
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		findLink(c, result)
		if *result != nil {
			return
		}
	}
}

func findParagraphText(n *html.Node) string {
	if n.Type == html.ElementNode && n.Data == "p" {
		return extractText(n)
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		if text := findParagraphText(c); text != "" {
			return text
		}
	}
	return ""
}

func getAttr(n *html.Node, key string) string {
	for _, a := range n.Attr {
		if a.Key == key {
			return a.Val
		}
	}
	return ""
}

func extractText(n *html.Node) string {
	if n.Type == html.TextNode {
		return n.Data
	}
	var sb strings.Builder
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		if c.Type == html.ElementNode && (c.Data == "script" || c.Data == "style") {
			continue
		}
		sb.WriteString(extractText(c))
	}
	return strings.TrimSpace(sb.String())
}

func init() {
	engine.Register("bing", New)
}
